restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -866,45 +766,46 @@ void addRowsToMap(Map<String, List<Row>> rowsByDoc, List<Row> rows) {
               new ArrayList<>(),
               false,
               null,
-              pageSize,
-              currentPageState);
-      List<Row> rows = page.left;
-      skipSeenRows(initialPagingState, rows);
+              paginator);
+      rows = paginator.maybeSkipRows(rows);
       addRowsToMap(rowsByDoc, rows);
-      firstRequest = false;
-    } while (rowsByDoc.keySet().size() <= limit && page.right != null);
+    } while (rowsByDoc.keySet().size() <= paginator.docPageSize && paginator.hasDbPageState());
 
     // Either we've reached the end of all rows in the collection, or we have enough rows
     // in memory to build the final result.
-    Set<String> docNames = rowsByDoc.keySet();
-    if (docNames.size() > limit) {
-      Iterator<Map.Entry<String, List<Row>>> iter = rowsByDoc.entrySet().iterator();
-      String lastIdSeen = null;
-      for (int i = 0; i < limit; i++) {
-        Map.Entry<String, List<Row>> e = iter.next();
-        List<Row> rows = new ArrayList<>();
-        for (Row row : e.getValue()) {
-          lastIdSeen = row.getString("key");
-          if (fields.isEmpty() || fields.contains(row.getString("p0"))) {
-            rows.add(row);
-          }
-        }
-        docsResult.set(e.getKey(), convertToJsonDoc(rows, false, db.treatBooleansAsNumeric()).left);
-      }
-      DocumentSearchPageState pageState = new DocumentSearchPageState(lastIdSeen, currentPageState);
-      return ImmutablePair.of(docsResult, pageState);
+    if (rowsByDoc.keySet().size() > paginator.docPageSize) {
+      AtomicReference<String> lastIdSeen = new AtomicReference<>();
```

Comment:
`MutableObject<String>` would also work here as you are not in parallel stream..